### PR TITLE
Updates for MOI 0.6

### LINKED ISSUE #63: Support MOI 0.6
Are there relevant issues we are waiting here and in MOI?
@IssamT still interested in doing it?

### PR DESCRIPTION
**Pull request by IssamT**
Tested with LQOI and with GLPK. Closes #63

### CODE REVIEW

File: REQUIRE [hunk 1]
```diff
@@ -1,3 +1,3 @@
 julia 0.6
-MathOptInterface 0.5 0.6
+MathOptInterface 0.5 0.7
```
Comment (odow) [inline REQUIRE]: You should bump the lower bound to 0.6

File: src/constraints.jl [hunk 2]
```diff
@@ -141,20 +131,24 @@ end
...
-    return haskey(model.constraint_names_rev, name)
-end
-function MOI.get(model::LinQuadOptimizer, ::Type{<:MOI.ConstraintIndex}, name::String)
-    return model.constraint_names_rev[name]
+function MOI.get(model::LinQuadOptimizer, ::Type{<:MOI.ConstraintIndex}, 
+                 name::String)
+    if haskey(model.constraint_names_rev, name)
+        return model.constraint_names_rev[name]
+    else
+        return nothing
```
Comment (odow) [inline src/constraints.jl]: I'm not sure that it should return `nothing`. Shouldn't it throw an error?

File: src/constraints.jl [hunk 3]
```diff
@@ -141,20 +131,24 @@ end
...
-end
-function MOI.get(model::LinQuadOptimizer, ::Type{MOI.ConstraintIndex{F,S}}, name::String) where F where S
-    model.constraint_names_rev[name]::MOI.ConstraintIndex{F,S}
+function MOI.get(model::LinQuadOptimizer, ::Type{MOI.ConstraintIndex{F,S}}, 
+                 name::String) where F where S
+    if haskey(model.constraint_names_rev, name) && 
+            isa(model.constraint_names_rev[name], MOI.ConstraintIndex{F,S})
+        return model.constraint_names_rev[name]::MOI.ConstraintIndex{F,S}
+    else
+        return nothing
```
Comment (odow) [inline src/constraints.jl]: Ditto here

File: src/objective.jl [hunk 4]
```diff
@@ -99,26 +99,24 @@ function MOI.supports(model::LinQuadOptimizer, ::MOI.ObjectiveFunction{F}) where
...
 function MOI.get(model::LinQuadOptimizer, ::MOI.ObjectiveFunction{MOI.SingleVariable})
-    if model.obj_type != MOI.SingleVariable
-        error("Cannot get SingleVariable objective function as the model has " *
-              " a $(model.obj_type) objective function.")
+    if model.obj_type != SingleVariableObjective
+        if VERSION >= v"0.7-"
+            throw(InexactError(:convert, SingleVariableObjective, model.obj_type))            
+        else
+            throw(InexactError())
+        end
```
Comment (odow) [inline src/objective.jl]: What was wrong with the current error?

File: src/variables.jl [hunk 5]
```diff
@@ -53,17 +52,19 @@ end
...
-function MOI.canget(model::LinQuadOptimizer, ::Type{MOI.VariableIndex}, name::String)
-    return haskey(model.variable_names_rev, name)
-end
+
 function MOI.get(model::LinQuadOptimizer, ::Type{MOI.VariableIndex}, name::String)
-    return model.variable_names_rev[name]
+    if haskey(model.variable_names_rev, name)
+        return model.variable_names_rev[name]
+    else
+        return nothing
```
Comment (odow) [inline src/variables.jl]: Ditto here

File: src/objective.jl (hunk 4 above)
Comment (IssamT) [reply src/objective.jl]: Just trying to be closer to MOI requirements. https://github.com/JuliaOpt/MathOptInterface.jl/blob/master/src/attributes.jl#L392-L399

File: src/constraints.jl (hunk 2 above)
Comment (IssamT) [reply src/constraints.jl]: It follows MOI requirements:
https://github.com/JuliaOpt/MathOptInterface.jl/blob/master/src/attributes.jl#L160-L163

Whether it's good or not is another question, and should be tackled on MOI level.

File: src/objective.jl [hunk 6]
```diff
@@ -130,36 +128,30 @@ function MOI.get(model::LinQuadOptimizer, ::MOI.ObjectiveFunction{Linear})
...
-    sizehint!(quadratic_terms, length(coefficients))
-    for (column, variable) in enumerate(model.variable_references)
-        for j in nzrange(Q, column)
-            row = rows[j]
-            push!(quadratic_terms,
-                  MOI.ScalarQuadraticTerm{Float64}(
-                      coefficients[j],
-                      model.variable_references[row],
-                      variable)
-            )
```
Comment (joaquimg) [inline src/objective.jl]: Why there is no more objective type?

Comment (IssamT) [reply src/objective.jl]: What do you mean?
Notice that I modified the behavior of this function according to the requirement of `MOI.canget(model::LinQuadOptimizer, ::MOI.ObjectiveFunction{Quad})`: an optimizer must return a proper quadratic function even when all the quadratic terms are zero (i.e. the function is actually linear or a single variable).

**Comment by IssamT** [comment]
2 LQOI tests are failing because of this

https://github.com/JuliaOpt/LinQuadOptInterface.jl/blob/master/test/contconic.jl#L27

which causes the failure of 

https://github.com/JuliaOpt/MathOptInterface.jl/blob/master/src/Test/contconic.jl#L213

and

https://github.com/JuliaOpt/MathOptInterface.jl/blob/master/src/Test/contconic.jl#L261

@joaquimg, is it possible to make the return status more precise ?

File: test/runtests.jl [hunk 7]
```diff
@@ -98,10 +98,10 @@ const LQOI = LinQuadOptInterface
         MOIT.lin2vtest(solver, config)
         set_lin2test_solutions!(solver)
         MOIT.lin2ftest(solver, config)
-        set_lin3test_solutions!(solver)
-        MOIT.lin3test(solver, config)
-        set_lin3test_solutions!(solver)
-        MOIT.lin4test(solver, config)
+        # set_lin3test_solutions!(solver)
```
Comment (IssamT) [inline test/runtests.jl]: I commented these to make sure the other tests are passing. We should solve this https://github.com/JuliaOpt/LinQuadOptInterface.jl/pull/64#issuecomment-418730742 and uncomment them before we merge

File: src/objective.jl (hunk 6 above)
Comment (joaquimg) [reply src/objective.jl]: ok!

**Comment by joaquimg** [comment]
Does anyone knows why the CI on 0.7 is having all those problems?

**Comment by tkoolen** [comment]
Would you mind tagging a release with MOI 0.5 support first? This would make package maintenance a lot easier for me, so I'd really appreciate it.

**Comment by odow** [comment]
Yes, good idea in a release. I can't seem to draft a new release on my phone, I'll do it tonight when I get back to my computer.

**Comment by joaquimg** [comment]
Yep, I could tag if its just that...

**Comment by joaquimg** [comment]
Can I tag master as 0.3.1 before merging this?
@tkoolen @odow

**Comment by joaquimg** [comment]
BTW @odow is okay with thus PR or there are further changes needed?

**Comment by odow** [comment]
@joaquimg I'm busy moving. Please tag :)

**Comment by mlubin** [comment]
0.3.1 is tagged, safe to merge?

**Pull request merged**